tests_pytorch/test_attacks.py

```diff
@@ -0,0 +1,330 @@
+        continue
+
+  def test_eps_iter_equals_zero(self):
+    for ord in self.ord_list:
+      try:
+        self.assertClose(
+            self.attack(
+                model_fn=self.model, x=self.x, eps=.5, ord=ord, nb_iter=10,
+                eps_iter=0),
+            self.x)
+      except NotImplementedError:
+        continue
+
+  def test_invalid_clips(self):
+    clip_min = .5
+    clip_max = -.5
+    for ord in self.ord_list:
+      try:
+        self.assertRaises(
+            AssertionError, self.attack, model_fn=self.model, x=self.x, eps=.1,
+            ord=ord, clip_min=clip_min, clip_max=clip_max, nb_iter=10,
+            eps_iter=.01)
+      except NotImplementedError:
+        continue
+
+  def test_adv_example_success_rate_linf(self):
+    # use normalized_x to make sure the same eps gives uniformly high attack
+    # success rate across randomized tests
+    self.help_adv_examples_success_rate(
+        ord=np.inf, **self.attack_param)
+
+  def test_targeted_adv_example_success_rate_linf(self):
+    self.help_targeted_adv_examples_success_rate(
+        ord=np.inf, **self.attack_param)
+
+  def test_adv_example_success_rate_l1(self):
+    try:
+      self.help_adv_examples_success_rate(
+          ord=1, **self.attack_param)
+    except NotImplementedError:
+      raise SkipTest()
+
+  def test_targeted_adv_example_success_rate_l1(self):
+    try:
+      self.help_targeted_adv_examples_success_rate(
+          ord=1, **self.attack_param)
+    except NotImplementedError:
+      raise SkipTest()
+
+  def test_adv_example_success_rate_l2(self):
+    self.help_adv_examples_success_rate(
+        ord=2, **self.attack_param)
+
+  def test_targeted_adv_example_success_rate_l2(self):
+    self.help_targeted_adv_examples_success_rate(
+        ord=2, **self.attack_param)
+
+  def test_do_not_reach_lp_boundary(self):
+    for ord in self.ord_list:
+      try:
+        x_adv = self.attack(
+            model_fn=self.model, x=self.normalized_x, eps=.5, nb_iter=10,
+            ord=ord, eps_iter=.01)
+      except NotImplementedError:
+        continue
+      if ord == np.inf:
+        delta, _ = torch.abs(x_adv - self.normalized_x).max(dim=1) 
+      elif ord == 1:
+        delta = torch.abs(x_adv - self.normalized_x).sum(dim=1)
+      elif ord == 2:
+        delta = torch.pow(x_adv - self.normalized_x, 2).sum(dim=1).pow(.5)
+      diff = torch.max(.5 - delta)
+      self.assertTrue(diff > .25)
+
+  def test_attack_strength(self):
+    x_adv = self.attack(
+        model_fn=self.model, x=self.normalized_x, eps=1.,
+        eps_iter=.05, ord=np.inf, clip_min=.5, clip_max=.7, nb_iter=5,
+        sanity_checks=False)
+    _, ori_label = self.model(self.normalized_x).max(1)
+    _, adv_label = self.model(x_adv).max(1)
+    adv_acc = adv_label.eq(ori_label).sum().to(torch.float)\
+        / self.normalized_x.size(0)
+    self.assertLess(adv_acc, .1)
+
+  def test_eps(self):
+    # test if the attack respects the norm constraint
```

Comment:
I was unsure about this when I left it there. My concern was that if someone reads test_attacks.py but not test_utils.py, she might think we haven't implemented any test on whether the attacker respects the given norm constraint. So I put a reminder in test_attacks.py making sure that whoever reads the code is clear that this test is implemented somewhere else. 

Do you think this is unnecessary? If so I can remove it. Or should I change it to just a comment instead of a dummy test?